### ISSUE TITLE
Job list is empty

### ISSUE THREAD

**Issue by fleimgruber** [kabouzeid/turm#17]
As a user, running `turm` shows the TUI with the 3 main panes, but without any jobs. No keyboard press has a visible effect, only q for quitting.

I compiled turm myself and we use Slurm 18.08. Is it maybe a compatibility issue?

**Comment by kabouzeid** [comment]
It just parses the output of squeue. So if squeue works, so should turm. Not sure what's going on.

**Comment by fleimgruber** [comment]
Can you give me a hint on how to best debug this from turm? I triple checked that squeue without args gives me a job list, but turm does not. Since I only have a CLI available I tried rust-gdb, but it's output interferes with the turm TUI.

**Comment by fleimgruber** [comment]
I am not intending to sound cheeky, but if there were automated tests shipped with turm, I could try running these on the SLURM system...

**Comment by kabouzeid** [comment]
would love to have tests, but then you have to somehow setup a clean slurm environemnt and start dummy jobs there. not sure how to best do that.

this is the part you need to debug:
https://github.com/kabouzeid/turm/blob/f104c7c646880f3881a99fa183ce5165cbf8c5b3/src/job_watcher.rs#L53-L133

**Comment by fleimgruber** [comment]
True, maybe this could provide a clean environment for testing? https://hub.docker.com/r/hpcnow/slurm_simulator

Failing that I could also see a set of test job definitions maintained here that could be used for very basic testing, e.g. a few sleep jobs that print to stdout so that at least parts of the UI are tested.

Regarding the part to debug: I do not yet have a CLI debugging setup for Rust. Another idea that came to mind: there is a feature of other Slurm TUIs to use SSH to connect to a Slurm host so the TUI would run locally and could then be more easily debugged, e.g. visual debugger in VS Code. Did you think about remote Slurm access? Do you have experience with SSH in Rust?

**Comment by kabouzeid** [comment]
You can use the remote SSH VS Code extension for running and debugging on the slurm host.

**Comment by fleimgruber** [comment]
Thanks for mentioning, a good idea! I tried debugging in VS Code which tells me to install LLDB extensions. After that LLDB fails with version `GLIBC_2.18' not found. Slurm is running on CentOS 7 which only has glibc 2.17. I think also other Rust dev tools need at least glibc 2.18? See also https://github.com/rust-lang/rust-analyzer/issues/4706.

**Comment by fleimgruber** [comment]
In the meantime, I would try "printf-debugging", but written to a file because stdout will be drawn with TUI main loop already. I have this template:

```rust
let path = "results.txt";
let mut output = File::create(path)?;
let job_command = ...
write!(output, "{}", job_command)
```

Could you provide guidance on what to insert at `...` from `jobs` to get the full `squeue` command that will be tried?

**Comment by kabouzeid** [comment]
Just debug print the `Command` with 

```rust
let cmd = Command::new("squeue") 
     .args(&self.squeue_args) 
     .arg("--array") 
     .arg("--noheader") 
     .arg("--Format") 
     .arg(&output_format)

println!("{:?}", cmd);
```

**Comment by fleimgruber** [comment]
For met it only works with

```rust
let cmd = Command::new("squeue")
      .args(&self.squeue_args)
      .arg("--array")
      .arg("--noheader")
      .arg("--Format")
      .arg(&output_format)
      .output();
println!("{:?}", cmd);
```

which prints a string with the expected comma-separated fields.

**Comment by fleimgruber** [comment]
Ok, I could further narrow it down to this check:
https://github.com/kabouzeid/turm/blob/f104c7c646880f3881a99fa183ce5165cbf8c5b3/src/job_watcher.rs#L67
which always evaluates to `true` so it always returns None.

**Comment by fleimgruber** [comment]
And the actual cause I think is that:
https://github.com/kabouzeid/turm/blob/f104c7c646880f3881a99fa183ce5165cbf8c5b3/src/job_watcher.rs#L65
does not split at `###turm###` because it is not included in the output of `squeue`.

**Comment by fleimgruber** [comment]
So as mentioned in OP, it actually is a compatibility issue with Slurm 18.08. Do you see another way to do the string post-processing? E.g. split on a tab or a certain amount of blanks instead of the `###turm###` sentinel.

**Comment by kabouzeid** [comment]
Thanks for tracking this down!

> Edit: I see now that the only way to parse the output is to not use the --noheader argument and look for the header column positions to correctly infer the field offsets for the actual output lines.

If someone implements this in a robust enough way, I would be willing to merge it. I won't have time to do this myself.

**Comment by fleimgruber** [comment]
I went ahead and implemented my suggested approach from https://github.com/kabouzeid/turm/issues/17#issuecomment-1768644298 in #20
